### ISSUE TITLE
Regression for type variables bound to literals

### ISSUE THREAD

**Issue by RobertCraigie** [microsoft/pyright#4457]
**Describe the bug**

This snippet worked prior to `1.1.289`

```py
from __future__ import annotations

from typing import Literal, TypeVar, Mapping

_T = TypeVar('_T', bound=Literal['country', 'city'])

def foo(by: list[_T], order: Mapping[_T, str] | None = None) -> None:
    ...

foo(['country'])
foo(['country'], order={'country': 'asc'})
```

Pyright seems to infer the `by` and `order` arguments to `list[str]` and `dict[str, str]` which is not correct.

```
foo.py
 foo.py:13:6 - error: Argument of type "list[str]" cannot be assigned to parameter "by" of type "list[_T@foo]" in function "foo"
    Type "Literal['country']" cannot be assigned to type "_T@foo" (reportGeneralTypeIssues)
  foo.py:14:6 - error: Argument of type "list[str]" cannot be assigned to parameter "by" of type "list[_T@foo]" in function "foo"
    Type "Literal['country']" cannot be assigned to type "_T@foo" (reportGeneralTypeIssues)
  foo.py:14:24 - error: Argument of type "dict[str, str]" cannot be assigned to parameter "order" of type "Mapping[_T@foo, str] | None" in function "foo"
    Type "dict[str, str]" cannot be assigned to type "Mapping[_T@foo, str] | None"
      "dict[str, str]" is incompatible with "Mapping[_T@foo, str]"
        TypeVar "_KT@Mapping" is invariant
          Type "str" cannot be assigned to type "_T@foo"
      Type cannot be assigned to type "None" (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations
```

**Expected behavior**

The given snippet should pass type checks.

**VS Code extension or command-line**

Command-line `v1.1.289`

**Comment by erictraut** [comment]
This was an intentional change, not a regression. It was needed to fix another issue in the constraint solver.

Mypy also doesn't accept the above code, so I thought it was unlikely that anyone was relying on the need to bind a type variable to a literal type.

Can you explain the scenario where a type variable bound to a literal (or a union of literals) is useful?

**Comment by RobertCraigie** [comment]
In my case it's used to represent the relationships between the arguments, e.g. you cannot specify a field in `order` that you haven't included in `by`.

```py
results = await Profile.prisma().group_by(
    ['country'],
    order={
        # city has to be included in the `by` argument to be valid
        'city': 'asc',
    }
)
```

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by RobertCraigie** [comment]
Amazing, thank you so much!

**Comment by erictraut** [comment]
This is included in pyright 1.1.290, which I just published. It will also be included in a future release of pylance.
